### PR TITLE
translator & tts

### PR DESCRIPTION
**Pull request by ParadoxSpace**
image
image
image
image
image

port a shit.

Ports handheld, ear, and omni translators. Handheld translators and ear translators translate any language that is not a sign language, a nonverbal one, a hivemind, or any of the machine_understands = false languages. Omni translators are capable of doing these machine_understands = false languages, but still doesn't do hiveminded/nonverbal/signlangs. 

Also a TTS device. Basically like the circuit one but simpler and easily available for mute characters, or characters who have troubles speaking properly, such as Security.

### CODE REVIEW

**Comment by ParadoxSpace** [comment]
big shoutout to my main man gemspizzle for pointing out how fucking bad i am

File: code/game/objects/items/devices/text_to_speech.dm [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+	if(user.incapacitated(INCAPACITATION_ALL)) //Are you in a state to actual use the device?
+		to_chat(user, "You cannot activate the device in your state.")
+		return
+
+	if(!named)
+		to_chat(user, "You input your name into the device.")
+		name = "[initial(name)] ([user.real_name])"
+		desc = "[initial(desc)] This one is assigned to [user.real_name]."
+		named = 1
+		/* //Another way of naming the device. Gives more freedom, but could lead to issues.
```
Comment (Arrow768) [inline code/game/objects/items/devices/text_to_speech.dm]: Dead Code. Remove it.

File: code/modules/client/preference_setup/loadout/loadout_utility.dm [hunk 2]
```diff
@@ -65,4 +65,14 @@
 	for(var/fanny in typesof(/obj/item/weapon/storage/belt/fannypack))
 		var/obj/item/weapon/storage/belt/fannypack/fanny_type = fanny
 		fannys[initial(fanny_type.name)] = fanny_type
-	gear_tweaks += new/datum/gear_tweak/path(sortAssoc(fannys))
\ No newline at end of file
+	gear_tweaks += new/datum/gear_tweak/path(sortAssoc(fannys))
+
+/datum/gear/utility/translator //riskiest PR I've ever made.
```
Comment (Arrow768) [inline code/modules/client/preference_setup/loadout/loadout_utility.dm]: This is not fit for the general loadout.
It has relatively high tech levels and gives a immediate advantage to the players in the game.
(Be it for the use in the science destructive analyzer or for its ability to understand most languages)

File: code/game/objects/items/devices/text_to_speech.dm [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+		device_name = copytext(sanitize(input(user, "What would you like to name your device? You must input a name before the device can be used.", "Name your device", "") as null|text),1,MAX_NAME_LEN)
+		name = "[initial(name)] - [device_name]"
+		named = 1
+		*/
+
+	var/message = sanitize(input(user,"Choose a message to relay to those around you.") as text|null)
+	if(message)
+		var/obj/item/device/text_to_speech/O = src
+		audible_message("\icon[O] \The [O.name] states, \"[message]\"")
```
Comment (Arrow768) [inline code/game/objects/items/devices/text_to_speech.dm]: The TTS device should use a universally understood language (i.e. TCB) and not output the message in a "magic" way that allows all mobs regardless of them speaking the language or not to understand said message.

**Comment by ParadoxSpace** [comment]
gg

File: code/modules/client/preference_setup/loadout/loadout_utility.dm (hunk 2 above)
Comment (ParadoxSpace) [reply code/modules/client/preference_setup/loadout/loadout_utility.dm]: it doesn't have high tech levels. it has origin_tech = list(TECH_DATA = 3, TECH_ENGINEERING = 3)

File: code/game/objects/items/devices/text_to_speech.dm (hunk 1 above)
Comment (ParadoxSpace) [reply code/game/objects/items/devices/text_to_speech.dm]: will do

File: code/game/objects/items/devices/text_to_speech.dm (hunk 3 above)
Comment (ParadoxSpace) [reply code/game/objects/items/devices/text_to_speech.dm]: will do

File: code/modules/client/preference_setup/loadout/loadout_utility.dm (hunk 2 above)
Comment (Arrow768) [reply code/modules/client/preference_setup/loadout/loadout_utility.dm]: A good point, but with the addition of universal translators the recorders should loose the ability to act as universal translators

Comment (ParadoxSpace) [reply code/modules/client/preference_setup/loadout/loadout_utility.dm]: that would make them useless for investigations (their primary function)

File: code/game/objects/items/devices/text_to_speech.dm (hunk 3 above)
Comment (ParadoxSpace) [reply code/game/objects/items/devices/text_to_speech.dm]: on second thought 2complex

File: code/modules/client/preference_setup/loadout/loadout_utility.dm (hunk 2 above)
Comment (Arrow768) [reply code/modules/client/preference_setup/loadout/loadout_utility.dm]: How so ? You can always ask another crew member to help you translate if you do not speak the language.
_Or the universal translator_

Comment (ParadoxSpace) [reply code/modules/client/preference_setup/loadout/loadout_utility.dm]: Another crew member cannot produce a transcript that is very hard to forge.

Comment (ParadoxSpace) [reply code/modules/client/preference_setup/loadout/loadout_utility.dm]: Also, furthermore, they only translate the basic languages that crew use, and not even anything with nonverbal or signlang. (And you can only use them when they're in your hand)

Comment (ParadoxSpace) [reply code/modules/client/preference_setup/loadout/loadout_utility.dm]: Again, this does not leave a transcript. It impacts the performance of both CCIA and detectives.

File: code/game/objects/items/devices/translator.dm [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	var/visual = 1		//If you need to see to get the message
+	var/audio = 0		//If you need to hear to get the message
+	var/omni = 0	//A terrible mistake. If the language translated doesn't need machine_understands
+	var/listening = 0
+	var/datum/language/langset
+
+/obj/item/device/universal_translator/attack_self(mob/user)
+	if(!listening) //Turning ON
+		langset = input(user,"Translate to which of your languages?","Language Selection") as null|anything in user.languages
+		if(langset && ((langset.flags & NONVERBAL) || (langset.flags & HIVEMIND) || (langset.flags & SIGNLANG) || (!langset.machine_understands)))
```
Comment (skull132) [inline code/game/objects/items/devices/translator.dm]:
```suggestion
		if(langset && ((langset.flags & (NONVERBAL|HIVEMIND|SIGNLANG)) || (!langset.machine_understands)))
```

**Pull request closed without merging**